### PR TITLE
VEGAS+: Numpy support and more tests

### PR DESCRIPTION
**Pull request by FHof**
Summary of changes:
* Numpy support for VEGAS
  * Replace torch. with autoray.numpy. operations
    After this commit it still only works with torch and should indirectly perform the same torch operations as before. I put these changes in an separate commit so that if a regression was introduced in some commit later, it is easier to find the bug.
  * Replace some numerical operations so that VEGAS+ works with both torch and numpy
  * Add dtype arguments so that it is possible to integrate with float32 precision with numpy
* A new VEGASMap._smooth_map method
* A small docstring fix in VEGAS.integrate
* More tests
  * Support other backends in vegas_test.py and use it to test with Numpy
  * Add special peak integrands to vegas_test.py to check that VEGAS+ works better than MonteCarlo for these integrands
  * Tests for VEGASMap and VEGASStratification

Relevant issue: #14

### CODE REVIEW

File: torchquad/integration/vegas_map.py [hunk 1]
```diff
@@ -67,120 +92,125 @@ def get_Jac(self, y):
...
+            self.delta_weights,
+        ) = self._smooth_map(self.weights, self.counts, self.alpha)
 
         for i in range(self.dim):  # Update per dim
             old_i = 0
             d_accu = 0
 
-            indices = torch.zeros(self.N_intervals - 1).long()
-            d_accu_i = torch.zeros(self.N_intervals - 1)
+            indices = [0] * (self.N_intervals - 1)
```
Comment (gomezzz) [inline torchquad/integration/vegas_map.py:210]: This is the thing that you mentioned where it is more efficient to stay with python datatypes, I think, right? Please add a comment here and below where you then use anp to explain this

File: torchquad/integration/vegas_stratification.py [hunk 2]
```diff
@@ -8,41 +8,58 @@ class VEGASStratification:
...
+            rng (RNG): Random number generator
+            backend (string): Numerical backend
+            dtype (backend dtype): dtype used for the calculations
             beta (float, optional): Beta parameter from VEGAS Enhanced. Defaults to 0.75.
         """
+        self.rng = rng
         self.dim = dim
         # stratification steps per dim, EQ 33
-        self.N_strat = math.floor((N_increment / 2.0) ** (1.0 / dim))
+        self.N_strat = int((N_increment / 2.0) ** (1.0 / dim))
```
Comment (gomezzz) [inline torchquad/integration/vegas_stratification.py:25]: I think you wanted to change this to 4, right?

**Review by gomezzz — CHANGES_REQUESTED** [review]
A few minor comments / questions, overall looking good 👍

File: torchquad/integration/vegas_map.py [hunk 3]
```diff
@@ -12,16 +13,24 @@ def __init__(self, dim, integration_domain, N_intervals=100, alpha=0.5) -> None:
 
         Args:
             dim (int): Dimensionality of the integrand.
-            integration_domain (list, optional): Integration domain, e.g. [[-1,1],[0,1]]. Defaults to [-1,1]^dim.
+            integration_domain (backend tensor): Integration domain.
```
[the remark this answers is not in the document]
Comment (FHof) [reply torchquad/integration/vegas_map.py:16]: I think the docstring was already wrong before my changes. There is no default value in the code, so I removed it from the docstring.
https://github.com/esa/torchquad/blob/d1e532e8986e5ded5ee7f371ad0071c5db1c2663/torchquad/integration/vegas_map.py#L10

File: torchquad/integration/vegas_map.py [hunk 4]
```diff
@@ -67,120 +92,125 @@ def get_Jac(self, y):
...
-
-            # EQ 20
-            self.delta_weights[dim] = self.summed_weights[dim] / self.N_intervals
+        dim, N_intervals = weights.shape
+        i_tmp = N_intervals - 2
+        d_tmp = anp.concatenate(
+            [
+                7.0 * weights[:, 0:1] + weights[:, 1:2],
+                weights[:, :-2] + 6.0 * weights[:, 1:-1] + weights[:, 2:],
+                weights[:, i_tmp : i_tmp + 1] + 7.0 * weights[:, i_tmp + 1 : i_tmp + 2],
```
[the remark this answers is not in the document]
Comment (FHof) [reply torchquad/integration/vegas_map.py:161]: Yes, the tensors need to have the same shape except on axis 1 so that they can be concatenated.

File: torchquad/integration/vegas_map.py [hunk 5]
```diff
@@ -67,120 +92,125 @@ def get_Jac(self, y):
...
     def update_map(
         self,
     ):
         """Update the adaptive map, Section II C."""
-        self._smooth_map()
+        (
+            self.smoothed_weights,
+            self.summed_weights,
+            self.delta_weights,
+        ) = self._smooth_map(self.weights, self.counts, self.alpha)
```
[the remark this answers is not in the document]
Comment (FHof) [reply torchquad/integration/vegas_map.py:204]: I haven't investigated this. 
In comparison to PyTorch and Numpy, Tensorflow and especially JAX do not support item assignment (see for example https://jax.readthedocs.io/en/latest/jax.ops.html), so I think JAX and Tensorflow don't support re-using existing memory unless the functions are compiled.

File: torchquad/integration/vegas_stratification.py (hunk 2 above)
Comment (FHof) [reply torchquad/integration/vegas_stratification.py:25]: Yes, I haven't added the changes to this PR yet.
I think in VegasFlow they calculate `int(N_increment / 2) / 2`, which approximately corresponds to division by 4,
and in the original implementation it's `(1 - neval_frac) * self.neval / 2.`, where neval_frac is some number between 0 and 1.
https://github.com/N3PDF/vegasflow/blob/21209c928d07c00ae4f789d03b83e518621f174a/src/vegasflow/vflowplus.py#L111-L112
https://github.com/gplepage/vegas/blob/11f67af45c8fc5c8c25060f18a61137274a64789/src/vegas/_vegas.pyx#L1095

Comment (FHof) [reply torchquad/integration/vegas_stratification.py]: I've added a commit to change it to 4, but the errors in the test are worse now. Previous errors:
```
[0.0, 0.00023080781786584048, 0.0007456416333443405, 0.008323833256319801, 1.5603112756120936, 0.010654779995036279, 9.560055281099267e-05, 1.670036573481193e-05, 0.0005694908455957659]
``` 
Errors with fewer hypercubes:
```
[0.0013054318169358758, 0.0017626261337132091, 0.0009283768196267639, 0.026356434890061564, 4.905033994889436, 0.018269307164842985, 0.0014304650819014775, 2.7983103974504875e-05, 0.0004786292093108546]
```

I think in VegasFlow they calculate `int(N_increment / 2) / 2`, which approximately corresponds to division by 4,
and in the original implementation it's `(1 - neval_frac) * self.neval / 2.`, where neval_frac is some number between 0 and 1.
https://github.com/N3PDF/vegasflow/blob/21209c928d07c00ae4f789d03b83e518621f174a/src/vegasflow/vflowplus.py#L111-L112
https://github.com/gplepage/vegas/blob/11f67af45c8fc5c8c25060f18a61137274a64789/src/vegas/_vegas.pyx#L1095

File: torchquad/integration/vegas_map.py (hunk 1 above)
Comment (FHof) [reply torchquad/integration/vegas_map.py:213]: I added the comments.

**Review by gomezzz — APPROVED** [review]
